base/compiler/inferencestate.jl

```diff
@@ -1116,3 +1121,74 @@ function get_max_methods_for_module(mod::Module)
     max_methods < 0 && return nothing
     return max_methods
 end
+
+"""
+    Future{T}
+
+Delayed return value for a value of type `T`, similar to RefValue{T}, but
+explicitly represents completed as a `Bool` rather than as `isdefined`.
+Set once with `f[] = v` and accessed with `f[]` afterwards.
+
+Can also be constructed with the `completed` flag value and a closure to
+produce `x`, as well as the additional arguments to avoid always capturing the
+same couple of values.
+"""
+mutable struct Future{T}
+    x::T
+    completed::Bool
+    Future{T}() where {T} = (f = new{T}(); f.completed = false; f)
+    Future{T}(x) where {T} = new{T}(x, true)
+    Future(x::T) where {T} = new{T}(x, true)
+end
+getindex(f::Future) = (@assert f.completed; f.x)
```

Comment:
`@assert` is not supposed to be used for checks thay ensure correctness of the operation, if I'm not mistaken?

Reply:
That is basically what this assert is for. It helps when developing to catch if something is wrong, but the correctness should be dealt with elsewhere (notably, `f.x` is usually an UndefRefException)